deltacat/compute/compactor/steps/materialize.py

```diff
@@ -101,69 +151,60 @@ def materialize(
         # single filter on average, and thus provides better average
         # performance than repeatedly filtering the table in dedupe task index
         # order
         dedupe_task_indices = chain.from_iterable(dedupe_task_idx_iter_tpl)
-        compacted_table = sc.append_dedupe_task_idx_col(
-            compacted_table,
+        pa_table = sc.append_dedupe_task_idx_col(
+            pa_table,
             dedupe_task_indices,
         )
         pa_sort_keys = [(sc._DEDUPE_TASK_IDX_COLUMN_NAME, "ascending")]
-        compacted_table = compacted_table.take(
-            pc.sort_indices(compacted_table, sort_keys=pa_sort_keys),
+        pa_table = pa_table.take(
+            pc.sort_indices(pa_table, sort_keys=pa_sort_keys),
         )
-        compacted_table = compacted_table.drop(
+        pa_table = pa_table.drop(
             [sc._DEDUPE_TASK_IDX_COLUMN_NAME]
         )
-        compacted_tables.append(compacted_table)
-
-    # TODO (pdames): save memory by writing output files eagerly whenever
-    #  len(compacted_table) >= max_records_per_output_file (but don't write
-    #  partial slices from the compacted_table remainder every time!)
-    compacted_table = pa.concat_tables(compacted_tables)
-    if compacted_file_content_type in DELIMITED_TEXT_CONTENT_TYPES:
-        # convert to pandas since pyarrow doesn't support custom delimiters
-        # and doesn't support utf-8 conversion of all types (e.g. Decimal128)
-        # TODO (pdames): compare performance to pandas-native materialize path
-        df = compacted_table.to_pandas(
-            split_blocks=True,
-            self_destruct=True,
-        )
-        del compacted_table
-        compacted_table = df
-    delta = deltacat_storage.stage_delta(
-        compacted_table,
-        partition,
-        max_records_per_entry=max_records_per_output_file,
-        content_type=compacted_file_content_type,
-    )
-    manifest = delta.manifest
-    manifest_records = manifest.meta.record_count
-    assert(manifest_records == len(compacted_table),
-           f"Unexpected Error: Materialized delta manifest record count "
-           f"({manifest_records}) does not equal compacted table record count "
-           f"({len(compacted_table)})")
-    materialize_result = MaterializeResult.of(
-        delta,
-        mat_bucket_index,
-        # TODO (pdames): Generalize WriteResult to contain in-memory-table-type
-        #  and in-memory-table-bytes instead of tight coupling to paBytes
-        PyArrowWriteResult.of(
-            len(manifest.entries),
-            TABLE_CLASS_TO_SIZE_FUNC[type(compacted_table)](compacted_table),
-            manifest.meta.content_length,
-            len(compacted_table),
-        ),
-    )
-    logger.info(f"Materialize result: {materialize_result}")
+
+        # Write manifests up to max_records_per_output_file
+        if compacted_tables and \
+                total_record_count + record_count > max_records_per_output_file:
```

Comment:
what if single delta manifest entry has more than max_records_per_output_file record?

Reply:
We raise an exception asking the client to increase the `records_per_compacted_file` parameter. This should generally be okay since this parameter is adjustable by the caller on a per-compaction job/round basis.

Reply:
got it.